src/mpi/errhan/errutil.c

```diff
@@ -1061,18 +1060,23 @@ int MPIR_Err_create_code_valist(int lastcode, int fatal, const char fcname[],
              * We do this here because we cannot both access the
              * user error code and pass the argp to vsnprintf_mpi . */
             if (specific_msg) {
-                const char *specific_fmt;
                 int specific_idx;
-                user_error_code = va_arg(Argp, int);
+                const char *specific_fmt;
                 specific_idx = FindSpecificMsgIndex(specific_msg);
                 if (specific_idx >= 0) {
                     specific_fmt = specific_err_msgs[specific_idx].long_name;
                 } else {
                     specific_fmt = specific_msg;
                 }
+                user_error_code = va_arg(Argp, int);
+                if (user_error_code & ERROR_DYN_MASK) {
+                    use_user_error_code = 1;
+                } else if (is_valid_error_class(user_error_code)) {
+                    error_class = MPIR_ERR_GET_CLASS(user_error_code);
+                }
```

Comment:
So if user callback return a specific code that falls in the `else` branch, we just overwrite? What if the code has a specific meaning to the user?

Is the motivation from mpi4py always inspecting a return code using e.g. `MPI_Error_string` and you can't tell if it is from a user callback? I think the proper way is for mpi4py to wrap a user callback return with a specific (or dynamic) error code before passing thru MPI. That way, you ensure raw user code never gets to MPI and you can design your own mechanism of preserving user code.

Reply:
> So if user callback return a specific code that falls in the `else` branch, we just overwrite?

Yes. Although the exact value will end up in the error string.

>  What if the code has a specific meaning to the user?

Then the user should rather register and return a dynamic error code.

> Is the motivation from mpi4py always inspecting a return code using e.g. `MPI_Error_string` and you can't tell if it is from a user callback? 

You have to understand that I do not do things just because of silly and frivolous mpi4py motivations.

To your specific question, no, I do not by default call `MPI_Error_string`. But that could eventually happen, if an uncaught exception is printed in the the traceback.

Checking every MPI call for errors and eventually querying the error string  is not something exclusive of mpi4py. PETSc also checks the return code of MPI calls, and gets the error string to embed that information in their own error code handling.

> I think the proper way is for mpi4py to wrap a user callback return with a specific (or dynamic) error code before passing thru MPI. That way, you ensure raw user code never gets to MPI and you can design your own mechanism of preserving user code.

That's what mpi4py already does, and has done that for years! Python user callbacks are executed in a `try/except` block, and any exception is translated to returning error code. Actually, I'm not even that exactly because you refused my interpretation of delete callbacks as destructors for attribute values.

Again, this is not about Python and mpi4py!!! The root of the problem is that an outer MPI routine returning a non-MPI error code not know to MPI is a bomb in user hands waiting to explode: there is no standard MPI way to tell whether the error code is a valid MPI error code, and calling `MPI_Error_string` on such error code is expected to trigger to an additional failure (that is, asking the error string of an unknown error code), thus swallowing the original point of failure.

Long story short: In my opinion, allowing MPI routines to return unknown MPI error codes is bad, simply because it does not compose well across third-party library boundaries. If libraryA creates a generalized request that ends up in the hands of libraryB calling `MPI_Wait` on it, and a grequest callback returns a non MPI error code, then the return value is a) meaningless to libraryB, and b) useless because an attempt to `MPI_Error_string` would either fail or return garbage. Therefore, non-MPI errors should not be let pass through. This is the heart of the problem I'm trying to address. Can we focus on that and forget about Python and mpi4py?

Reply:
> The root of the problem is that an outer MPI routine returning a non-MPI error code not know to MPI is a bomb in user hands waiting to explode

The "bomb" originates from the user and returns back to the same user to consume. It is just passing thru MPI. It may be a bomb, or may be part of user's meaningful design. So my philosophy is we handle it with care and return back to user and trust user know what they are doing. There are obviously the case when users don't know what they are doing, but since we don't know that, too eager to help may end up messing up "expert" users.

Reply:
> The "bomb" originates from the user and returns back to the same user to consume.

Not necessarily. I already mentioned libraryA returning a non-MPI error code, and libraryB having to deal with that unknown error code. What's your proposal for such scenario?

Reply:
Where does this non-MPI error code originate from, libraryA or libraryB?

Reply:
I already gave an example involving generalized requests. LibraryA returns a generalized request, let say the query function (implemented in LibraryA) may eventually returns a non-MPI error code. The generalized request is passed to LibraryB, and eventually LibrayyB will wait for the request. The waiting triggers the invocation of the query callback (in LbraryA land), that then returns a non MPI error code that gets propagated to the Wait call in (in LibraryB land). 

How is LibaryB supposed to deal with the non-MPI return code generated in LibraryA? Of course, you may argue that LibraryA should document things (or fix their code to do it the MPI way), and LibraryB should be prepared for the LibraryA behavior. But sometimes that's very hard to achieve in practice, simply because modern software is composed of many layers involving many independent libraries.

Reply:
In this case the non-MPI error originates in LibraryA (from the query function) and is passing through libraryB. LibraryB should manage to pass the error back to libraryA since libraryA may depend on that special non-MPI error. How libraryB do that maybe complicated but it depends on the interface between libraryA and libraryB. You may argue that a `MPI_Wait` call is innocent. But since libraryB is waiting on a generalized request from libraryA, there must be an API design for that.